plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+  -- but w/o the root node and only uses the Var's Unique as the node id
+  -- OPTIMIZE: by using AdjacencyIntMap
+  depGraph :: AM.AdjacencyMap PLC.Unique
+  depGraph = AM.gmap (\case Variable u -> u; _ -> error "just for completion")
+             $ AM.removeVertex Root -- we remove Root because we do not care about it
+             $ runTermDeps pir
+
+  -- | the dependency graph as before, but with datatype-related nodes merged/reduced under a single node per datatypebind a, see 'principal'.
+  reducedDepGraph :: AM.AdjacencyMap PLC.Unique
+  reducedDepGraph = M.foldr (\ (_,_,b,_) accGraph ->
+                               let ids = bindingIds b
+                               in case ids of
+                                 [_] -> accGraph -- optimization to avoid an O(n) op. if there is nothing to merge
+                                 _   ->  AM.mergeVertices (`S.member` S.fromList ids) (b^.principal) accGraph
+
+                            ) depGraph rhsTable
+
+  -- |take the strongly-connected components of the reduced dep graph, because it may contain loops (introduced by the LetRecs)
+  -- topologically sort these sccs, since we rely on linear (sorted) scoping in our 'genLets' code generation
+  topSortedSccs :: [S.Set PLC.Unique]
+  topSortedSccs =
+    let allLets = M.keysSet rhsTable in
+    -- we remove lambda sccs because we are not interested in them
+    -- for optimization, we just check one of the elements if it is a let unique. lambdas appear only by themselves.
+    filter (\ scc-> S.findMin scc `S.member` allLets)
+    . fmap AMN.vertexSet -- we are not interested in graph structure anymore
+    . fromRight (error "Cycle detected in the depgraph. This shouldn't happen in the first place.") $ AM.topSort $ AM.scc reducedDepGraph
+
+  -- | Groups a given set of lets into one or more multilets and wraps these multilets around a given term.
+  -- The grouping is done through the strongly-connected components
+  -- The input lets is not sorted w.r.t. linear scoping, so this function uses the topological-sort of these SCCs,
+  -- to figure out the correct (dependend/linear) order in which to generate these new multilets.
+  --
+  -- The resulting term is wrapped with linear-scope-sorted LetRecs and LetNonRecs (interspersed between each other because letnonrec and letrec syntax cannot be combined)
+  -- Example: `let {i = e, ...} in let rec {j = e, ...} in let rec {...} in let {...} in ..... in originalTerm`
+  genLets :: S.Set PLC.Unique -- ^ all the let identifiers to wrap around this term
+           -> FloatData -- ^ the remaining data to be floated
+           -> Term tyname name uni a                           -- ^ the term to be wrapped
+           -> State [S.Set PLC.Unique] (Term tyname name uni a)                           -- ^ the final wrapped term
+  genLets lets restDepthTable t = do
+    (hereSccs, restSccs) <- gets $ splitSccs lets
+    put $! restSccs
```

Comment:
There isn't much point in forcing a list to WHNF, especially given the fact that you fully force `hereSccs` on the next line, i.e. `splitSccs` also gets forced.